### ISSUE TITLE
Add `doctest` support to run docstring example automatically

### ISSUE THREAD

**Issue by antalszava** [PennyLaneAI/pennylane#2907]
### Feature details

It would be good to add `doctest` support. See https://github.com/PennyLaneAI/pennylane/pull/705.

### How important would you say this feature is?

1: Not important. Would be nice to have.

**Comment by timmysilv** [comment]
note that I ran doctest with today's master (and a config that just does `import pennylane as qml`) and saw ~5000 failures, so I don't believe that simply enabling doctest will be an option.

That said, I will leave this issue open for discoverability/discussion

**Comment by josh146** [comment]
@timmysilv out of curiosity do you know what the main failure reason was? Something that shouldn't work (e.g., docstrings with external hardware device examples) or things that _should_ work?

**Comment by timmysilv** [comment]
some heavy-hitting NameErrors (the number is the number of occurrences):
```
  83     NameError: name 'qscript' is not defined
  95     NameError: name 'tape' is not defined
  98     NameError: name 'torch' is not defined
 110     NameError: name 'dev' is not defined
 160     NameError: name 'op2' is not defined
 160     NameError: name 'op3' is not defined
 182     NameError: name 'U' is not defined
 213     NameError: name 'circuit' is not defined
 590     NameError: name 'np' is not defined

I also saw 104 instances of `TypeError: 'module' object is not callable`

**Comment by timmysilv** [comment]
alternate idea: make it opt-in with `IPython` directives instead. you can configure imports using [ipython_execlines](https://github.com/pandas-dev/pandas/blob/888b6bc1bb3fdf203dda64e50336db4eb9b1e778/doc/source/conf.py#L473-L475) in doc/conf.py, then you can change code blocks to use it with the [IPython Sphinx Directive](https://ipython.readthedocs.io/en/stable/sphinxext.html). The main con of this IMO is that it shows IPython-style output which can be kinda ugly (see the linked doc for examples).

There might be something similar for pycon (eg. [sphinx-autorun](https://pypi.org/project/sphinx-autorun/), though idk how maintained it is) which would ofc be great.

**Comment by astralcai** [comment]
@andrijapau related to what you're doing now for hack week.

**Comment by andrijapau** [comment]
> [@andrijapau](https://github.com/andrijapau) related to what you're doing now for hack week.

Yup, trying this out here: https://github.com/PennyLaneAI/pennylane/pull/8324/

**Comment by andrijapau** [comment]
We now have `doctest` support in our CI using `sybil`. Closing this issue. 😄 

We are currently working on cleaning up our documentation examples so that they all pass.
